src/VisualStudio/Next/Remote/ServiceHubRemoteHostClient.JsonRpcSnapshot.cs

```diff
@@ -0,0 +1,159 @@
+
+            // communication channel related to service information
+            private readonly JsonRpcClient _serviceClient;
+
+            // close connection when cancellation has raised
+            private readonly CancellationTokenRegistration _cancellationRegistration;
+
+            public JsonRpcSession(
+                ChecksumScope snapshot,
+                Stream snapshotStream,
+                object callbackTarget,
+                Stream serviceStream,
+                CancellationToken cancellationToken) :
+                base(snapshot, cancellationToken)
+            {
+                _snapshotClient = new SnapshotJsonRpcClient(this, snapshotStream);
+                _serviceClient = new ServiceJsonRpcClient(serviceStream, callbackTarget);
+
+                // dispose session when cancellation has raised
+                _cancellationRegistration = CancellationToken.Register(Dispose);
+            }
+
+            public override Task InvokeAsync(string targetName, params object[] arguments)
+            {
+                CancellationToken.ThrowIfCancellationRequested();
+
+                return _serviceClient.InvokeAsync(targetName, arguments.Concat(ChecksumScope.SolutionChecksum.Checksum.ToArray()).ToArray());
+            }
+
+            public override Task<T> InvokeAsync<T>(string targetName, params object[] arguments)
+            {
+                CancellationToken.ThrowIfCancellationRequested();
+
+                return _serviceClient.InvokeAsync<T>(targetName, arguments.Concat(ChecksumScope.SolutionChecksum.Checksum.ToArray()).ToArray());
+            }
+
+            public override Task InvokeAsync(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync)
+            {
+                CancellationToken.ThrowIfCancellationRequested();
+
+                return _serviceClient.InvokeAsync(targetName, arguments.Concat(ChecksumScope.SolutionChecksum.Checksum.ToArray()).ToArray(), funcWithDirectStreamAsync, CancellationToken);
+            }
+
+            public override Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync)
+            {
+                CancellationToken.ThrowIfCancellationRequested();
+
+                return _serviceClient.InvokeAsync<T>(targetName, arguments.Concat(ChecksumScope.SolutionChecksum.Checksum.ToArray()).ToArray(), funcWithDirectStreamAsync, CancellationToken);
+            }
+
+            protected override void OnDisposed()
+            {
+                // dispose cancellation registration
+                _cancellationRegistration.Dispose();
+
+                // dispose service and snapshot channels
+                _serviceClient.Dispose();
+                _snapshotClient.Dispose();
+            }
+
+            private class ServiceJsonRpcClient : JsonRpcClient
```

Comment:
Can you make each class specify what they're there for?  also 'Service' is very ague.  Which service is this?